Sparse io

tombena opened the pull request:
Specify percentage of IOs to be skipped using option P.

Commit:
Metadata files are now hidden

MD files are now hidden in the directory
where the test is launched.

Help message for -i modified

Commit:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	yapio.c

Commit:
Sparse IO

Introducing new option in recipe
P{percentage} can be added to a
scenario, it skips that percentage
of IOs

Commit:
Sparse IO

Introducing new option in recipe
P{percentage} can be added to a
scenario, it skips that percentage
of IOs

tombena committed:
Merge branch 'master' into sparse_io

Nicolas-Triquenaux commented on yapio.c:
```diff
@@ -725,6 +732,33 @@ yapio_parse_test_recipe(const char *recipe_str)
+                yapio_print_help(YAPIO_EXIT_ERR);
+            }
+            /* find where percentage stops */
+            int end = i + 2;
```
you should use a variable that to store that +2 as it is used in many places, if it chances there is a hight chance of forgetting to update every one. Furthermore it is not very meaning full, and can help to understand your operations if you store that into a variable with a proper name

tombena replied:
len("P.") = 2, i'll make it clearer in the next version

Nicolas-Triquenaux commented on yapio.c:
```diff
@@ -725,6 +732,33 @@ yapio_parse_test_recipe(const char *recipe_str)
+            ytc->ytc_sparse_io = percent * ytg->ytg_num_blks_per_rank;
+            int j;
+            for (j = i + 2; j < end; j++)
+                ytc->ytc_sparse_io = ytc->ytc_sparse_io/10;
```
I fail to see the value of that code. But i might be wrong

 You already have the length of percentage, i'll bet (end - i +2) so why not do that in a one liner ? i would say something line the following ?
 ytc->ytc_sparse_io = (percent * ytg->ytg_num_blks_per_rank) / ((end- i +2) * 10) ; 

As it is int, you'll not be bothered with the floating point approximation error. If i was the case, refrain yourself from doing lots of operations as the approximation propagates.

Furthermore you are missing one division by 10 in the current setup of you loop.

tombena replied:
I think it's more like *10^len(percent)*. I don't see how to avoid the successive divisions.

If P.0100 and num_blks_per_rank = 100 we should skip 1% of 100 IOs so that's 1 IO.
We get (variable) percent = 100 and len(percent) = 4
therefore: IO skipped = 100 *  100 / 10^4 = 1

Nicolas-Triquenaux commented on yapio.c:
```diff
@@ -1349,6 +1384,49 @@ yapio_blk_md_t *
+    printf("numops = %d\n", num_ops);
+
+    /* TODO need to YAPIO_FREE elsewhere */
+    bool *skipped_ios = YAPIO_MALLOC(num_ops * sizeof(bool));
```
I hope you did it

00pauln00 commented on yapio.c:
```diff
@@ -228,6 +228,7 @@ typedef struct yapio_blk_metadata
     int    ybm_writer_rank;     //rank who has last written this block
     int    ybm_owner_rank_fpp;  //fpp only, file to which contents belong
     size_t ybm_blk_number;      //number of the block - should not change!
+    bool   ybm_not_hole;        //block is empty
```
Adding a new member here increases the structure size by 8 bytes - which is considerable for this structure (since there's one per block).  Can you please figure out a way to squeeze this into a single bit possibly by using a bitfield?  ybm_blk_number should be fine with only 63 bits.

00pauln00 commented on yapio.c:
```diff
@@ -725,6 +732,33 @@ yapio_parse_test_recipe(const char *recipe_str)
+            ytc->ytc_sparse_io = percent * ytg->ytg_num_blks_per_rank;
+            int j;
+            for (j = i + 2; j < end; j++)
+                ytc->ytc_sparse_io = ytc->ytc_sparse_io/10;
```
whitespace around the "/" operator.

00pauln00 commented on yapio.c:
```diff
@@ -1349,6 +1384,49 @@ yapio_blk_md_t *
+    printf("numops = %d\n", num_ops);
+
+    /* TODO need to YAPIO_FREE elsewhere */
+    bool *skipped_ios = YAPIO_MALLOC(num_ops * sizeof(bool));
```
This is going to allocate a lot of memory when the number of blocks is large.  Since we already have array items for each block is this really needed?

ntriquenaux replied:
+1

Nicolas-Triquenaux replied:
:+1:

Commit:
Merge branch 'sparse_io' of https://github.com/tombena/yapio into sparse_io

Conflicts:
	yapio.c

Commit:
Simplify sparse IO

Skipped IO array is no longer used
but ytc_blk_number is reduced to 62
bits so we can add bits for skip IO
and holes

Commit:
Simplify sparse IO

Skipped IO array is no longer used
but ytc_blk_number is reduced to 62
bits so we can add bits for skip IO
and holes

Commit:
Merge branch 'sparse_io' of https://github.com/tombena/yapio into sparse_io
